File: src/main/java/waterpower/common/network/PacketTileEntity.java
package waterpower.common.network;

import io.netty.buffer.ByteBuf;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import net.minecraft.nbt.CompressedStreamTools;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;
import waterpower.WaterPower;
import waterpower.common.block.tileentity.TileEntityBase;

public class PacketTileEntity implements IMessage, IMessageHandler<PacketTileEntity, IMessage> {

    public long pos;
    public TileEntityBase base;
    public NBTTagCompound tag;

    public PacketTileEntity() {
    }

    public PacketTileEntity(TileEntityBase entity) {
        this();
        pos = entity.getPos().toLong();
        base = entity;
    }

    @Override
    public IMessage onMessage(PacketTileEntity message, MessageContext ctx) {
        World world = WaterPower.getWorld();
        TileEntity te = world.getTileEntity(BlockPos.fromLong(message.pos));
        if (te instanceof TileEntityBase)
            message.base = (TileEntityBase) te;
        else
            message.base = null;
        if (message.base == null)
            return null;

        message.base.readPacketData(message.tag);

        return null;
    }

    @Override
    public void fromBytes(ByteBuf buf) {
        pos = buf.readLong();

        int len = buf.readInt();
        byte[] data = new byte[len];
        buf.readBytes(data);
        try {
            tag = CompressedStreamTools.read(new DataInputStream(new ByteArrayInputStream(data)));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void toBytes(ByteBuf buf) {
        buf.writeLong(pos);
        if (tag == null) {
            tag = new NBTTagCompound();
            base.writePacketData(tag);
        }
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        try {
            CompressedStreamTools.write(tag, dos);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        byte[] data = baos.toByteArray();
        buf.writeInt(data.length);
        buf.writeBytes(data);
    }

}

File: src/main/kotlin/waterpower/common/network/PacketTileEntity.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.network

import io.netty.buffer.ByteBuf
import net.minecraft.nbt.CompressedStreamTools
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.relauncher.Side
import waterpower.annotations.Message
import waterpower.common.block.tile.TileEntityBase
import waterpower.util.getWorld
import java.io.*

@Message(Side.CLIENT)
class PacketTileEntity() : Packet<PacketTileEntity>() {

    var pos: Long = 0
    var base: TileEntityBase? = null
    var tag: NBTTagCompound? = null

    constructor(entity: TileEntityBase) : this() {
        pos = entity.pos.toLong()
        base = entity
    }

    override fun onMessage(message: PacketTileEntity, ctx: MessageContext): IMessage? {
        val world = getWorld()
        val te = world!!.getTileEntity(BlockPos.fromLong(message.pos))
        if (te is TileEntityBase)
            message.base = te
        else
            message.base = null
        if (message.base == null || message.tag == null)
            return null

        message.base!!.readPacketData(message.tag as NBTTagCompound)

        return null
    }

    override fun fromBytes(buf: ByteBuf) {
        pos = buf.readLong()

        val len = buf.readInt()
        val data = ByteArray(len)
        buf.readBytes(data)
        try {
            tag = CompressedStreamTools.read(DataInputStream(ByteArrayInputStream(data)))
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

    }

    override fun toBytes(buf: ByteBuf) {
        buf.writeLong(pos)
        if (tag == null) {
            tag = NBTTagCompound()
            base!!.writePacketData(tag as NBTTagCompound)
        }
        val baos = ByteArrayOutputStream()
        val dos = DataOutputStream(baos)
        try {
            CompressedStreamTools.write(tag, dos)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        val data = baos.toByteArray()
        buf.writeInt(data.size)
        buf.writeBytes(data)
    }

}